controllers/resources/lease.go

```diff
@@ -0,0 +1,249 @@
+		expectedExpiry := now.Add(leaseExpectedDuration)
+		actualExpiry := l.Spec.RenewTime.Add(time.Second * time.Duration(int(*l.Spec.LeaseDurationSeconds)))
+		if actualExpiry.Before(expectedExpiry) {
+			err := m.commonLeaseManager.RequestLease(ctx, node, leaseExpectedDuration+LeaseBuffer)
+			if err != nil {
+				m.log.Error(err, "couldn't renew lease", "lease name", l.Name)
+				return 0, err
+			}
+		}
+		return leaseExpectedDuration, nil
+	}
+}
+
+func (m *nhcLeaseManager) InvalidateLease(ctx context.Context, remediationCR *unstructured.Unstructured) error {
+	node := &v1.Node{}
+	if err := m.client.Get(ctx, client.ObjectKey{Name: remediationCR.GetName()}, node); err != nil {
+		m.log.Error(err, "failed to get node", "node name", remediationCR.GetName())
+		return err
+	}
+
+	err := m.commonLeaseManager.InvalidateLease(ctx, node)
+	if err != nil {
+		m.log.Error(err, "failed to invalidate lease", "node name", remediationCR.GetName())
+		return err
+	}
+	return nil
+}
+
+func (m *nhcLeaseManager) getLeaseDurationForRemediation(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
+	if timeout := m.getTimeoutForRemediation(remediationCR, nhc); timeout != 0 {
+		return timeout
+	}
+	return DefaultLeaseDuration
+}
+
+func (m *nhcLeaseManager) getTimeoutForRemediation(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
+	var leaseDuration time.Duration
+	remediationKind := remediationCR.GetKind()
+	for _, esRemediation := range nhc.Spec.EscalatingRemediations {
+		if strings.TrimSuffix(esRemediation.RemediationTemplate.Kind, "Template") == remediationKind {
+			leaseDuration = esRemediation.Timeout.Duration
+			break
+		}
+	}
+	return leaseDuration
+}
+
+func (m *nhcLeaseManager) isLeaseOverdue(l *coordv1.Lease, nhc *remediationv1alpha1.NodeHealthCheck, remediationCR *unstructured.Unstructured) (bool, error) {
+	if l.Spec.AcquireTime == nil {
+		err := fmt.Errorf("lease Spec.AcquireTime is nil")
+		m.log.Error(err, "lease Spec.AcquireTime is nil", "lease name", l.Name)
+		return false, err
+	}
+
+	isLeaseOverdue := time.Now().After(m.calcLeaseExpiration(l, remediationCR, nhc))
+	return isLeaseOverdue, nil
+}
+
+func (m *nhcLeaseManager) calcLeaseExpiration(l *coordv1.Lease, remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Time {
+	leaseDuration := m.getLeaseDurationForRemediation(remediationCR, nhc)
+	prevRemediationsDuration := m.sumPrevRemediationsDuration(remediationCR, nhc)
+	return l.Spec.AcquireTime.Add(time.Duration(maxTimesToExtendLease+1 /*1 is offsetting the lease creation*/)*leaseDuration + prevRemediationsDuration)
+}
+
+func (m *nhcLeaseManager) isRemediationsExist(remediationCrs []unstructured.Unstructured) bool {
+	return len(remediationCrs) > 0
+}
+
+func (m *nhcLeaseManager) logManageLeaseChanges(originalRequeue time.Duration, updatedRequeue time.Duration, originalErr error, updatedErr error) {
```

Comment:
I couldn't find this being used. Did I miss something?

Reply:
No, good catch this is a leftover that was used in the old pattern where Reconcile iterate on nodes and Manage each  of them.